src/Data/Filters/UserSiteMatch.php

```diff
@@ -45,7 +45,8 @@ public function filter(\User $user, \Loris\Data\DataInstance $resource) : bool
             // If the Resource belongs to multiple CenterIDs, the user can
             // access the data if the user is part of any of those centers.
             foreach ($resource->getCenterIDs() as $site) {
-                if ($user->hasCenter($site)) {
+                $center = new \CenterID($site);
```

Comment:
This isn't correct. It's inside of a `instanceof \LORIS\StdyEntities\MultiSiteHaver` check. MultiSiteHaver is defined as

```
interface MultiSiteHaver
{
    /**
     * Return true if the entity is accessible by the
     * user.
     *
     * @return \CenterID[]
     */
    public function getCenterIDs() : iterable;
}
```

`$site` in the foreach loop should already be a centerid

Reply:
Hi @driusan The issue here seem to be variables types. `hasCenter()` is expecting and object of type `Center.` while **$resource->getCenterIDs()** is not returning it.

Reply:
What is the `$resource` that is not returning it and using the filter? The docblock for the interface is explicit that it's supposed to be returning CenterIDs (the only reason the php return type isn't is because php doesn't directly support that annotation.)

Reply:
According to the test done I would said `$resource` is not returning an object of type `CenterID`. When "manually" typecasted as in this PR it does take it.

Reply:
This is what shown at the error log before the fix. 
![image](https://user-images.githubusercontent.com/37309344/139701840-6bee30b2-592a-498a-9d7d-503fb5fa992e.png)

Reply:
What is the class type of `$resource`? The interface that it's known to be within the if statement is guaranteed to return CenterIDs according to the interface contract. If `getCenterIDs()` from something that is defined as a MultiSiteHaver is returning something other than `CenterID[]` it's a bug with that class, not places that use the interface.

Reply:
It seems there is not a "easy" way of solving this. The main issue seems to be due the PHP typeless default design. It seems we cannot have an array of Objects (at least not a classic array).

This post explains a bit more (dated for a couple of years but still current it seems).
https://stackoverflow.com/questions/40693469/function-return-type-hinting-for-an-array-of-objects-in-php7/40693938

We could try to make ourselves the class "array ofCenterIDs". I think we don't' have something similar in LORIS so far (a custom class array) and not sure the possible implications: it seems PHP array functions like` in_array()` will not work (at least not in a classic way again).

I see this kind change more as an architectural one, not sure if convenient/possible before the next release in a couple of weeks. Open to suggestions we could also discuss it  a bit more during the LORIS meeting.

Reply:
php can not, but phan (and most tooling) can follow the docblock typing when it's properly defined in a comment such as:

```
     * @return \CenterID[]
```

This should not involve changing the architecture or any code outside of user accounts, user accounts just needs to return a CenterID when it says it's going to.

See #7814 and #7815 for similar fixes to the issue_tracker and battery_manager

Reply:
I have been trying to implement the solution given in #7815, but somehow I'm getting the same (similar issues). I think I'm missing something. Sorry, seems will need help with this one.